src/osd/ECBackend.cc

```diff
@@ -1762,6 +1763,9 @@ void ECBackend::check_op(Op *op)
     dout(10) << __func__ << " Completing " << *op << dendl;
     writing.pop_front();
     tid_to_op_map.erase(op->tid);
+    bufferlist hbl;
+    op->hoid.encode(hbl);
```

Comment:
As I mentioned before, just index the cache on hobject_t.  It's probably not a good idea to need to run ::encode(hobject_t) on every cache access.  Also, bufferlist::c_str() *does not* null terminate the string (I know, it's not consistent with the stl).  Even if it did, there is no guarantee that the byte stream resulting from ::encode does not itself contain nulls!  You are implicitly calling std::string(char *) here (since set_evict takes a string by value and the std::string(char *) constructor isn't marked explicit).  That will take the pointer returned from hbl and copy up to the first null it finds creating a new std::string instance wrapping the newly allocated bytes.  This may not be (and almost certainly won't be since the encoding contains uint values which would have null bytes) the whole string returned by ::encode, and would overrun the buffer if the buffer didn't contain nulls.